Fix stats calculator [part 4]

arhimondr opened the pull request.

rschlussel commented on presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java:
```diff
@@ -160,11 +160,86 @@ protected PlanNodeStatsEstimate visitNotExpression(NotExpression node, Void cont
             }
-            PlanNodeStatsEstimate sourceStats = process(node.getValue());
-            if (sourceStats.isOutputRowCountUnknown()) {
+            PlanNodeStatsEstimate source = process(node.getValue());
```
I don't like the name source here. To me it means the same thing as input, which is confusing. maybe 'negatedFilter'?

arhimondr replied:
Aggree. Renamed `source` to `negatedFilter` everywhere in this method

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java:
```diff
@@ -20,81 +20,6 @@
-        double leftValuesPerDistinctValue = leftRowCount * (1.0 - leftStats.getNullsFraction()) / leftStats.getDistinctValuesCount();
-        double rightValuesPerDistinctValue = rightRowCount * (1.0 - rightStats.getNullsFraction()) / rightStats.getDistinctValuesCount();
-        if (leftValuesPerDistinctValue <= rightValuesPerDistinctValue) {
-            // right values cover all left values for corresponding distinct values
```
I don't see that comment in new code. Is the logic preserved?

arhimondr replied:
Yes, the logic is preserved. I just lost the comment.

sopel39 replied:
Actually I don't think its preserved for logical binary expression.

arhimondr replied:
It is not preserved for `OR`, as the NDV logic for or is different.

arhimondr replied:
For OR expression this condition is always true, assuming that `row_count / (disjunct_1_ndvs + disjunct_2_ndvs)` is always `<= row_count / conjunct_1_2_ndvs`.

That is not always true for `NOT` though. Because although `original_ndv` is always `<=` than `negated_predicate_ndvs`, the original `row_count` is >= `negated_predicate_row_count`

arhimondr replied:
I'm trying to remember why, but i didn't add this check to the `OR` estimation as it seemed to be extra.

arhimondr replied:
>  it seemed to be extra.

But that i mean that the check will always be `true`.

sopel39 replied:
It isn't extra. This came from customer case. Lets have expression: 'E1 or R2` and:
```
R1=rows(E1) , R2=rows(E2), where intersect(R1, R2) = 0
```
Then if `rows(E1 AND E2)=R1` this check is needed to get remove NDVs for which there are no rows.

Similarly if `rows(E1 AND E2)=R1/2 (half of rows for each NDVs)` then this check is needed to preserve NDVs for which all rows were not subtracted.

arhimondr replied:
Could you please explain why the range merge must be under this check?

arhimondr replied:
My thoughts about estimating NDV's for OR were next

NDV_AND - number of distinct values that are matched by both predicates

NDV_E1 - number of distinct values that are matched by E1
NDV_E2 - number of distinct values that are matched by E2

NDV_E1_ONLY - number of distinct values that are matched only by E1
NDV_E2_ONLY - number of distinct values that are matched only by E2

NDV_E1 = NDV_E1_ONLY + NDV_AND
NDV_E2 = NDV_E2_ONLY + NDV_AND

NDV_E1 + NDV_E2 = NDV_E1_ONLY + NDV_E2_ONLY + 2xNDV_AND

So it looks like we should always subtract `NDV_AND`  to remove the `2x` multiplier.

Same for number of rows.

For `NOT` expression the situation is different. The number of distinct values must be negated, and not subtracted.

If number of input values is N, and number of input values negated is N / 3, than the result should be 2/3 * N, but only if rows per NDV is less in source than in negated result, because of the reasons you have explained.

sopel39 replied:
`OR` implementation is in principle: https://en.wikipedia.org/wiki/Inclusion%E2%80%93exclusion_principle.
However, there are many rows per NDV and depending on filter only some rows might be filtered. For instance, if you have filter expression: `the_column_that_interest_me < 5` then for each NDV all or none rows will be filtered out. If you have expression `some_other_column < 5` then only some rows (from `the_column_that_interest_me` column) will be filtered out per NDV.

> For NOT expression the situation is different. The number of distinct values must be "heuristically" negated, and not subtracted.

This is what `subtractSubset(inputStats, process(node.getValue());` would do.

sopel39 replied:
> Could you please explain why the range merge must be under this check?

If no NDVs are removed then original (left) range is preserved. If some NDVs are removed we could adjust the range. There is some faulty logic in `StatisticRange#subtract` that tries to do that, but I would simply remove it.

arhimondr replied:
Maybe i'm being really slow, but in case the `IF` check is applied - the result can be incorrect. Consider following situation:

```
INITIAL ROWS COUNT = 400
INITIAL NDV = 2

P1 SELECTED ROWS = 300
P2 SELECTED ROWS = 3

P1 SELECTED NDV = 2
P2 SELECTED NDV = 1

OR (SUM) ROWS = 303
OR (SUM) NDV = 3
OR (SUM) ROWS PER NDV = 101

AND (INTERSECTION) ROWS = 3
AND (INTERSECTION NDV) = 1
AND (INTERSECTION) ROWS PER NDV = 3

IF CHECKING:

`OR (SUM) ROWS PER NDV` IS GREATER THAN `AND (INTERSECTION) ROWS PER NDV`
RESULT NDV = `OR (SUM) NDV = 3`, THAT IS GREATER THAN INITIAL NDV, NOT CORRECT

IF NOT CHECKING:

RESULT DV = (OR (SUM) NDV = 3) - (AND (INTERSECTION NDV) = 1) = 2 - CORRECT
```

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java:
```diff
@@ -51,20 +51,20 @@
 
 import static com.facebook.presto.cost.ComparisonStatsCalculator.estimateExpressionToExpressionComparison;
```
I'm not a big fan of this change. Can we work out a proper semantics (it really should be doable) for `subtractColumnStats` instead of simply removing it and having non-obvious code duplicated? Maybe the method had semi-defined semantics but after this change we have an implicit semantics for rather complex logic.

The advantage of dedicated `subtractColumnStats` would be also that it semantics can be properly described in Javadoc. For stats code this is not alwyas obvious from code. I also worry that if there comes a need, there will be 3rd and 4th implementation of column stats subtracting (as people will think they need special handling and there isn't stats toolset), which is not desirable.

arhimondr replied:
Let me explain why did i decided to inline it. That method was used only for `OR` and for `NOT` estimates. The `OR` estimation involves both, summing stats and than subtracting `AND` stats. It seems that it would be much easier to follow that code if no higher level stats method were involved. In other words when doing both, `summing` and `substracting` inline. The code seems to be more transparent, and easier to follow

e.g.:

- setting low-high only taking into account both disjunts separately: https://github.com/prestodb/presto/pull/11973/files#diff-2c8e7079274d1e50a57e44770854b2c5R324
- setting avg size to the input avg size: https://github.com/prestodb/presto/pull/11973/files#diff-2c8e7079274d1e50a57e44770854b2c5R322
- setting ndvs in a more explicit manner that is easier to follow: https://github.com/prestodb/presto/pull/11973/files#diff-2c8e7079274d1e50a57e44770854b2c5R328

Now since the code for `OR` got inlined i wasn't sure that we need a separate "abstract" method that effective is used just to estimate the negation. For me it is also easier to read. As you explicitly assuming that what you are doing - is negating stats. Not subtracting them. 

Thoughts?

findepi replied:
IMHO there is no such concept as "stats subtraction". You can subtract numbers, but "10 NDV - 15 NDV" does not make sense, in general.
For `NOT` we want to take "complement" which is a special case of "subtraction" (and still, may be different than just numerical subtraction of each stat property). Do we have other places where exactly such a "complement" is being calculated? Do we do this for outer joins?

sopel39 replied:
> IMHO there is no such concept as "stats subtraction".

There is a concept of multiset (e.g: rows) subtraction and those sets are described by stats.

> You can subtract numbers, but "10 NDV - 15 NDV" does not make sense, in general.

It does with additional assumptions. We do something similar for join when we assume that NDVs match.

Anyaway, I propsed subtract implementation in comment.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java:
```diff
@@ -222,12 +297,61 @@ private PlanNodeStatsEstimate estimateLogicalOr(Expression left, Expression righ
+                return process(FALSE_LITERAL);
+            }
+
+            input.getSymbolsWithKnownStatistics().forEach(symbol -> {
```
What if we have methods as:

```suggestion
/**
 * Subtracts stats that are assumed to be subset of left stats
 */
PlanNodeStatsEstimate subtractSubset(PlanNodeStatsEstimate left, PlanNodeStatsEstimate right) {
  PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder();
  double outputRowCount = left.getOutputRowCount() - right.getOutputRowCount();
  verify(outputRowCount >= 0, "outputRowCount must be greater than or equal to zero: %s", outputRowCount);
  result.setOutputRowCount(outputRowCount);

  left.getSymbolsWithKnownStatistics().forEach(symbol -> {
    SymbolStatsEstimate leftSymbolStats = left.getSymbolStatistics(symbol);
    SymbolStatsEstimate rightSymbolStats = right.getSymbolStatistics(symbol);
    
    SymbolStatsEstimate.Builder newSymbolStats = SymbolStatsEstimate.builder();
    
    newSymbolStats.setLowValue(minNonNan(leftSymbolStats.getLowValue(), rightSymbolStats.getLowValue()));
    newSymbolStats.setHighValue(maxNonNan(leftSymbolStats.getHighValue(), rightSymbolStats.getHighValue()));
    
    double newDistinctValues = leftSymbolStats.getDistinctValuesCount() - rightSymbolStats.getDistinctValuesCount();
    
    double leftNumberOfNulls = leftSymbolStats.getNullsFraction() * leftEstimate.getOutputRowCount();
    double rightNumberOfNulls = rightSymbolStats.getNullsFraction() * rightEstimate.getOutputRowCount();
     
    double newNumberOfNulls = leftNumberOfNulls - rightNumberOfNulls;
    double newNullsFraction = min(newNumberOfNulls, outputRowCount) / outputRowCount;
     
    newSymbolStats.setNullsFraction(newNullsFraction);
    result.addSymbolStatistics(symbol, newSymbolStats.build());
  }
  
  return result.build();
}

PlanNodeStatsEstimate cap(PlanNodeStatsEstimate input, PlanNodeStatsEstimate cap) {
  PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder();
  double outputRowCount = min(input.getOutputRowCount(), cap.getOutputRowCount());
  result.setOutputRowCount(outputRowCount);
        
  input.getSymbolsWithKnownStatistics().forEach(symbol -> {
    SymbolStatsEstimate inputSymbolStats = input.getSymbolStatistics(symbol);
    SymbolStatsEstimate capSymbolStats = cap.getSymbolStatistics(symbol);
    
    SymbolStatsEstimate.Builder newSymbolStats = SymbolStatsEstimate.builder();
    
    // TODO: this should be capped too
    newSymbolStats.setLowValue(inputSymbolStats.getLowValue());
    newSymbolStats.setHighValue(inputSymbolStats.getHighValue());
    
    double newDistinctValues = minNonNan(inputSymbolStats.getDistinctValuesCount() , capSymbolStats.getDistinctValuesCount());
    
    double inputNumberOfNulls = inputSymbolStats.getNullsFraction() * inputEstimate.getOutputRowCount();
    double capNumberOfNulls = capSymbolStats.getNullsFraction() * capEstimate.getOutputRowCount();
     
    double newNumberOfNulls = minNonNan(inputNumberOfNulls, capNumberOfNulls);
    double newNullsFraction = min(newNumberOfNulls, outputRowCount) / outputRowCount;
     
    newSymbolStats.setNullsFraction(newNullsFraction);
    result.addSymbolStatistics(symbol, newSymbolStats.build());
  }
  
  return result.build();
}
```

Then for `OR` you simply do:
```
cap(subtractSubset(addStats(leftStats, rightStats), andStats), inputStats);
```

I think its more modular and easier to grasp. What do you think?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java:
```diff
@@ -160,11 +160,86 @@ protected PlanNodeStatsEstimate visitNotExpression(NotExpression node, Void cont
             }
-            PlanNodeStatsEstimate sourceStats = process(node.getValue());
-            if (sourceStats.isOutputRowCountUnknown()) {
+            PlanNodeStatsEstimate negated = process(node.getValue());
```
We could express `NOT` as:
```
subtractSubset(inputStats, process(node.getValue());
```
see my other comment

arhimondr commented:
First commit is copied from the https://github.com/prestodb/presto/pull/12027

sopel39 commented:
> I didn't find an easy way of extracting smaller methods when fixing the bugs and preserving the semantics. Also i added comprehensive unit tests for those two methods.

I did a quick shot on that: https://github.com/prestodb/presto/pull/12060 and I manged to extract some well defined methods. Feel free to do whatever you want with it.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java:
```diff
@@ -14,86 +14,196 @@
+        double supersetRowCount = superset.getOutputRowCount();
+        double subsetRowCount = subset.getOutputRowCount();
+        double outputRowCount = supersetRowCount - subsetRowCount;
+        verify(outputRowCount >= 0, "outputRowCount must be greater than or equal to zero: %s", outputRowCount);
```
we shouldn't fail here as stats are not perfect, we should just cap to 0

arhimondr replied:
That's was one of the reason why i wanted to have this inlined. When estimating `NOT` statement we know for sure that the negated stats shouldn't be bigger than the input stats, as the negated stats is a strict subset of the input. Having explicit assertions improved readability and makes it easier to understand what the code does.

sopel39 replied:
> When estimating NOT statement we know for sure that the negated stats shouldn't be bigger than the input stats, as the negated stats is a strict subset of the input

Double math is not perfect. It could be that `subset` rows equals to `superset + 0.0000001` rows. Therefore it cannot be assumed that stats will always be within constraints envelope. A lot of your recent refactors do just that (they make calculator more forgiving on case of bogus stats).  It seems we should be consistent here.

arhimondr replied:
> Double math is not perfect. It could be that subset rows equals to superset + 0.0000001 rows.

I think it makes sense. Let me change it.

> Why inlining is required?

The idea was that for the NOT estimate all the assertions always true, when for OR estimate (or in some other place where this method could be used) the assertions may start failing. But since we are going to replace assertion with capping - it is no longer the case.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java:
```diff
@@ -14,86 +14,196 @@
+            double supersetNullsCount = supersetSymbolStats.getNullsFraction() * supersetRowCount;
+            double subsetNullsCount = subsetSymbolStats.getNullsFraction() * subsetRowCount;
+            double newNullsCount = supersetNullsCount - subsetNullsCount;
+            verify(isNaN(newNullsCount) || newNullsCount >= 0, "newNullsCount must greater than or equal to zero: %s", newNullsCount);
```
ditto, we shouldn't fail here

arhimondr replied:
ditto

arhimondr replied:
removed assertion

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java:
```diff
@@ -14,86 +14,196 @@
+            // distinctValuesCount
+            double supersetDistinctValues = supersetSymbolStats.getDistinctValuesCount();
+            double subsetDistinctValues = subsetSymbolStats.getDistinctValuesCount();
+            double newDistinctValuesCount;
```
this can be simplified, see: https://github.com/prestodb/presto/pull/12060/commits/1d487a5e213128c6c572a2420ccd5a6820fd361f#diff-59fa6c2264eb72e0587e7e77ce155e59R75

arhimondr replied:
I'm trying to avoid division by zero. In the simplified version division by zero may happen.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java:
```diff
@@ -14,86 +14,196 @@
+                    newDistinctValuesCount = supersetDistinctValues;
+                }
+            }
+            verify(isNaN(newDistinctValuesCount) || newDistinctValuesCount >= 0, "newDistinctValuesCount must be greater or equal than zero: %s", newDistinctValuesCount);
```
ditto, we should fail here, but we should add 0 check, see: https://github.com/prestodb/presto/pull/12060/commits/1d487a5e213128c6c572a2420ccd5a6820fd361f#diff-59fa6c2264eb72e0587e7e77ce155e59R90

arhimondr replied:
The same explanation as here: https://github.com/prestodb/presto/pull/11973#discussion_r241064493

sopel39 replied:
Please not forget about 0 check

arhimondr replied:
removed assertion

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java:
```diff
@@ -14,86 +14,196 @@
+            verify(isNaN(newDistinctValuesCount) || newDistinctValuesCount >= 0, "newDistinctValuesCount must be greater or equal than zero: %s", newDistinctValuesCount);
+            newSymbolStats.setDistinctValuesCount(newDistinctValuesCount);
+
+            // range
```
just preserve superset range here (see: https://github.com/prestodb/presto/pull/12060/commits/1d487a5e213128c6c572a2420ccd5a6820fd361f#diff-59fa6c2264eb72e0587e7e77ce155e59R95). When trimming ranges we should account for NDVs density, but that's a whole different story out of scope for this PR

arhimondr replied:
Why should we preserve the input range if we can do better for negation? There's even an explicit test that covers that case: https://github.com/prestodb/presto/commit/1d487a5e213128c6c572a2420ccd5a6820fd361f#diff-9dbc32793a40a33af0bc1e8018d6b346R302

sopel39 replied:
Few reasons:
1. If the number of row per distinct value ("density check") in that particular range (that we subtract) is not enough, then `superset` range should be preserved.
2. If the `superset` has `N` distinct values in intersected range and `subset` has `M` distinct values in intersected range, then `superset` range should be preserved.

arhimondr replied:
I don't understand that. If full range for some column `A` is `[0, 5]`. Negated range is `[3,5]`. That means that all the values filtered by negated predicate are within `[3,5]`. That means that the only values left are within `[0, 3]` range. Why should it be conditional, and be based on the density check?

sopel39 replied:
Let me show an example:

> If the number of row per distinct value ("density check") in that particular range (intersection that we subtract) is not enough, then superset range should be preserved.

```
superset=[1,1,2,2,3,3,4,4], range(superset)=[1-4]
subset=[3,4], range(subset)=[3-4]
superset - subset=[1,1,2,2,3,4], range(superset - subset)=[1-4]
```

> If the superset has N distinct values in intersected range and subset has M distinct values in intersected range where N > M, then superset range should be preserved.

```
superset = [1,1,2,2,3.99999,3.99999,4,4], range(superset)=[1-4]
subset=[2,2,4,4], range(subset)=[2-3.99999]
superset - subset=[1,1,3.99999,3.99999], range(superset - subset)=[1-3.99999]
```

If you want to have better range negation, you should pushdown NOT to leaf predicates.

arhimondr replied:
Yes, you are right. I keep forgetting that range may be filtered by some other predicate on some other column. That actually makes sense. So it feels like we should remove the range negation altogether.

> If you want to have better range negation, you should pushdown NOT to leaf predicates.

What do you mean by that?

sopel39 replied:
> What do you mean by that?

I you have expression `NOT(a < 0 OR a > 10)` you will get better range estimate if you rewrite the expression to: `a >= 0 AND a <= 10`. This is because then we avoid `OR` whatsoever.

> So it feels like we should remove the range negation altogether.

I will take a shot at that (to account issues that I've mentioned) when this PR lands if you don't mind.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java:
```diff
@@ -14,86 +14,196 @@
     }
 
-    private static PlanNodeStatsEstimate differenceInStatsWithRangeStrategy(PlanNodeStatsEstimate left, PlanNodeStatsEstimate right, RangeSubtractionStrategy strategy)
+    public static PlanNodeStatsEstimate computeSymmetricDifferenceStats(
```
This method is a combo. One cannot deduce what it does from the method name (it is in fact Inclusion–exclusion `OR` implementation). It can be splited into: `cap` https://github.com/prestodb/presto/pull/12060/commits/1d487a5e213128c6c572a2420ccd5a6820fd361f#diff-59fa6c2264eb72e0587e7e77ce155e59R151, 
`addNonOverlapping` https://github.com/prestodb/presto/pull/12060/commits/1d487a5e213128c6c572a2420ccd5a6820fd361f#diff-59fa6c2264eb72e0587e7e77ce155e59R107 and `subtractSubset` (already exists). Each with well defined semantics.

arhimondr replied:
Symetric difference is a standard set operation: https://en.wikipedia.org/wiki/Symmetric_difference

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java:
```diff
@@ -14,86 +14,196 @@
+            double supersetDistinctValuesCount = supersetSymbolStats.getDistinctValuesCount();
+            if (range.getDistinctValuesCount() > supersetDistinctValuesCount) {
+                newSymbolStats.setDistinctValuesCount(supersetDistinctValuesCount);
+            }
```
there should be check for `newDistinctValueCount==0`.

arhimondr replied:
This is responsibility of normalization to set the range to empty, and the nulls fraction to `1` in case of `0` distinct values. We are not duplicating it anywhere else: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/cost/SymbolStatsEstimate.java#L74, https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/cost/StatsNormalizer.java#L122

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java:
```diff
@@ -14,86 +14,196 @@
+            StatisticRange range = leftSymbolStats.statisticRange().addAndSumDistinctValues(rightSymbolStats.statisticRange());
+            newSymbolStats.setStatisticsRange(range);
+            double supersetDistinctValuesCount = supersetSymbolStats.getDistinctValuesCount();
+            if (range.getDistinctValuesCount() > supersetDistinctValuesCount) {
```
there should be check for nan in this if, see https://github.com/prestodb/presto/pull/12060/commits/a86d6dcfb0461d35a9eb4c399dfcec78e1199f0a#diff-59fa6c2264eb72e0587e7e77ce155e59R138

arhimondr replied:
From my perspective, If something under `NOT` cannot estimate NDVs, the negation code shouldn't preserve the original NDV's.

arhimondr commented:
@sopel39 

Given https://github.com/prestodb/presto/pull/12060#issuecomment-446636221 and https://github.com/prestodb/presto/pull/12060#issuecomment-446685615 we would like to stick with the current implementation, if you don't mind. The main reasons for that is that the current implementation is easier to read, and it offers more (range negation for `NOT`).

CC: @rschlussel

arhimondr commented:
@sopel39 Since we decided not to do the range negation, i agree that `computeSymmetricDifferenceStats` can be replaced with 

```
return capStats(
                    subtractStats(
                            addStatsAndSumDistinctValues(leftEstimate, rightEstimate),
                            andEstimate),
                    input);
```

I removed the `computeSymmetricDifferenceStats` method, and added `capStats` method.

The only difference now between this, and your proposal - is that the density check is applied unconditionally for both, `OR` and `NOT`. 

CC: @rschlussel

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java:
```diff
@@ -222,12 +218,16 @@ private PlanNodeStatsEstimate estimateLogicalOr(Expression left, Expression righ
-            PlanNodeStatsEstimate sumEstimate = addStatsAndSumDistinctValues(leftEstimate, rightEstimate);
-            return differenceInNonRangeStats(sumEstimate, logicalAndEstimate);
+
+            return capStats(
```
nice! Thank you. Really appreciate that and I think it's for the good

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java:
```diff
@@ -14,86 +14,130 @@
 
-    @FunctionalInterface
-    private interface RangeSubtractionStrategy
+    public static PlanNodeStatsEstimate subtractStats(PlanNodeStatsEstimate superset, PlanNodeStatsEstimate subset)
```
Please make name more expressive, e.g: `subtractSubsetStats` and add Javadoc.

arhimondr replied:
Any ideas how javadoc should look like without replicating the code logic?

sopel39 replied:
Maybe:
```
    /**
     * Subtracts subset stats from supersets stats. It is assumed that each NDV from subset
     * has a matching NDV in superset.
     */
```
but it's a nit

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/StatisticRange.java:
```diff
@@ -173,25 +173,6 @@ public StatisticRange addAndCollapseDistinctValues(StatisticRange other)
         return new StatisticRange(minExcludeNaN(low, other.low), maxExcludeNaN(high, other.high), newDistinctValues);
     }
 
-    public StatisticRange subtract(StatisticRange rightRange)
```
+1

sopel39 commented on presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java:
```diff
@@ -118,27 +119,36 @@ public static PlanNodeStatsEstimate addStatsAndCollapseDistinctValues(PlanNodeSt
 
     private static PlanNodeStatsEstimate addStats(PlanNodeStatsEstimate left, PlanNodeStatsEstimate right, RangeAdditionStrategy strategy)
```
We should remove `strategy` from this method and call it as we talked: `addNonOverlapping` (or similarly). This can be a separate PR though. 

`addAndSumDistinctValues` and `addAndCollapseDistinctValues` should just be `addNonOverlapping`.
While `addAndMaxDistinctValues` should just be `addOverlapping`.

I could do that if you don't mind.

arhimondr replied:
I don't mind. But let's do this as a separate PR.

arhimondr commented:
merged
